storage/storage_image.go

```diff
@@ -0,0 +1,569 @@
+		Layers:         make(map[string][]string),
+		BlobData:       make(map[string][]byte),
+		SignatureSizes: []int{},
+	}
+	return &image, nil
+}
+
+func (s storageImageSource) Reference() types.ImageReference {
+	return s.imageRef
+}
+
+func (s storageImageDestination) Reference() types.ImageReference {
+	return s.imageRef
+}
+
+func (s storageImageSource) Close() {
+}
+
+func (s storageImageDestination) Close() {
+}
+
+func (s storageImageDestination) ShouldCompressLayers() bool {
+	// The storage layer automatically decompresses layers as part of
+	// applying them, so callers shouldn't bother doing that.
+	return true
+}
+
+// PutBlob is used to both store filesystem layers and binary data that is part
+// of the image.  Filesystem layers are assumed to be imported in order, as
+// that is required by some of the underlying storage drivers.
+func (s *storageImageDestination) PutBlob(stream io.Reader, blobinfo types.BlobInfo) (types.BlobInfo, error) {
+	blobSize := int64(-1)
+	digest := blobinfo.Digest
+	errorBlobInfo := types.BlobInfo{
+		Digest: "",
+		Size:   -1,
+	}
+	// Try to read an initial snippet of the blob.
+	header := make([]byte, 10240)
+	n, err := stream.Read(header)
+	if err != nil && err != io.EOF {
+		return errorBlobInfo, err
+	}
+	// Set up to read the whole blob (the initial snippet, plus the rest)
+	// while digesting it with either sha256, or the passed-in digest, if
+	// one was specified.
+	hasher := ddigest.SHA256.New()
+	if d := ddigest.Digest(digest); d.Validate() == nil {
+		if a := d.Algorithm(); a.Available() {
+			hasher = a.New()
+		}
+	}
+	hash := ""
+	counter := ioutils.NewWriteCounter(hasher.Hash())
+	defragmented := io.MultiReader(bytes.NewBuffer(header[:n]), stream)
+	multi := io.TeeReader(defragmented, counter)
+	if (n > 0) && archive.IsArchive(header[:n]) {
+		// It's a filesystem layer.  If it's not the first one in the
+		// image, we assume that the most recently added layer is its
+		// parent.
+		parentLayer := ""
+		if len(s.BlobList) > 0 {
+			for _, blob := range s.BlobList {
+				if layerList, ok := s.Layers[blob]; ok {
+					parentLayer = layerList[len(layerList)-1]
+				}
+			}
+		}
+		// If we have an expected content digest, generate a layer ID
+		// based on the parent's ID and the expected content digest.
+		id := ""
+		if ddigest.Digest(digest).Validate() == nil {
+			id = ddigest.SHA256.FromBytes([]byte(parentLayer + digest)).Hex()
+		}
+		// Attempt to create the identified layer and import its contents.
+		layer, uncompressedSize, err := s.imageRef.transport.store.PutLayer(id, parentLayer, nil, "", true, multi)
+		if err != nil && err != storage.ErrDuplicateID {
+			logrus.Debugf("error importing layer blob %q as %q: %v", blobinfo.Digest, id, err)
+			return errorBlobInfo, err
+		}
+		if err == storage.ErrDuplicateID {
+			// We specified an ID, and there's already a layer with
+			// the same ID.  Drain the input so that we can look at
+			// its length and digest.
```

Comment:
It’s absolutely not a _requirement_ to detect duplicate blobs; it would be perfectly fine for `GetExistingBlobInfo` to always return “unknown, make a new copy”. OTOH the docker/distribution server does detect the existence duplicate blobs across images, at least within the same registry; this is not theoretical and not redundant with the “duplicated layer digest” check within a single `copy.Image`.